cmd/snap-strace-shim/snap-strace-shim.c

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (C) 2025 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/ptrace.h>
+#include <unistd.h>
+
+#include "../libsnap-confine-private/utils.h"
+
+int main(int argc, char **argv) {
+    if (sc_is_debug_enabled()) {
+        for (int i = 0; i < argc; i++) {
+            fprintf(stderr, "-%s-\n", argv[i]);
+        }
+    }
+    if (argc < 2) {
+        fprintf(stderr, "missing a command to execute");
+        abort();
+    }
+
+    /* echo signal STOP to parent so that it knows when to attach strace */
+    raise(SIGSTOP);
+
+    const char *executable = argv[1];
```

Comment:
FWIW, a better way to detect execv() failure here is different, you need to inherit a pipe from the parent process, and set it CLOEXEC before execv().

If execv() returns you write a byte to the pipe.

Parent process just needs to poll() the pipe for being able to read and gets notified when execv() is succesfull - it's closed, or when it fails (a byte is available to read).

You can also write actual data to the pipe like an error message and read it from the parent.